---
title: "Bulk RNA sequencing for temporal profiling in repetitive mild traumatic brain injury"
author: "Alyssa Pybus"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, fig.align="center")
```


This project is organized into four separate scripts: data pre-processing/wrangling (Wrangle.R), a function to run analysis results on a given subset of the data (Sample Subset.R), the main analysis (22131-02 Analysis.R), and weighted gene coexpression network analysis (WGCNA.R). 


## Wrangle.R

Wrangle.R produces the following outputs for later use in the analysis:

- R data set of the filtered raw count matrix, metadata in matching order, and a data frame of the gene names and IDs
- R data set of sample protein values matrix, index of analyte columns within the protein data frame, greek character labels, and palettes for time and injury number
- R data get of gene sets for use in GSVA
- R data set of normalized, variance-stabilization gene expression


### Set-up

Clear out global environment, set working directory to current file location (where your current script is saved), and load necessary packages. You'll need to install any packages you don't already have that are listed in p_load().

```{r setup_wrangle}
rm(list=ls())
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
pacman::p_load(tidyverse)
```

### Transcript data preparation

#### Load the metadata

The metadata will later be sorted to match the order of samples from the data, so this variable has the suffix .us for "unsorted".
```{r wrangle_metadata_load}
# load in experiment data and metadata
metadata.us = readxl::read_excel("Data/22131-02 metadata.xlsx") %>%
  mutate(Injury = factor(Injury,levels=c("Sham","1xCHI","pre-3xCHI","3xCHI","pre-5xCHI","5xCHI")))

glimpse(metadata.us)
```

#### Load in and filter the count data

We use filter criteria to match recommendations for use in Weighted Gene Co-expression Network Analysis (WGCNA): keep genes where at least 90% have a raw count higher than 10.

Reference: <https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html> 

```{r wrangle_data_load}
count.in = rio::import("Data/22131-02-gene-count-all.csv")
gene.df.uf = count.in[,2:3]  # unfiltered gene names (non-unique) and EnsemblIDs (unique)
count.uf = count.in[,4:42]  # unfiltered gene counts
colnames(count.uf) = metadata.us$Sample
rownames(count.uf) = gene.df.uf$EnsemblID

# filter criteria as recommended on WGCNA FAQ (genes where 90% of samples have count >10)
keepInd=which(rowSums(count.uf>10) > {ncol(count.uf)*.9})
count.us = count.uf[keepInd,]   # unsorted columns (samples), filtered genes (rows)
gene.df = gene.df.uf[keepInd,]  # filtered genes (rows)

# sort by AbsoluteTime
metadata = dplyr::arrange(metadata.us,AbsoluteTime)  # sort sample metadata by ascending time point since first injury
sort_ind = match(metadata$Sample,metadata.us$Sample)
count <- count.us[,sort_ind]  # sort count matrix columns by ascending time point since first injury (match metadata tibble)

save(count,metadata,gene.df,file="R Data/genecounts.rda")  # save to access later

glimpse(count)
glimpse(metadata)
glimpse(gene.df)
```



### Protein data preparation

#### Load the protein data, filter, and subtract background

Load in the protein data, filter to just female frontal cortex samples (bulk RNAseq was conducted on female sematomotor cortex, adjacent to the frontal cortex where proteins were measured in the same animals), and subtract background values.

```{r wrangle_protein_load}
# load protein data
df_in <- readxl::read_excel("Data/DoD TBI Acute All.xlsx") %>%
  dplyr::arrange(Region,AbsoluteTime) %>%
  dplyr::filter(Sex == "F") %>%  # Female samples only (RNAseq was only on female samples)
  dplyr::filter(Region == "FC")

# define the analytes to treat as numeric variables by setting first and last in range
first_analyte="GFAP"
last_analyte="TNF-a"
analyte_range={which(colnames(df_in)==first_analyte)}:{which(colnames(df_in)==last_analyte)}

#create array of average background for each analyte
df_b <- df_in %>%
  filter(Sample == "background") %>%
  dplyr::select(all_of(analyte_range)) %>%
  colMeans()

#create data frame of sample data
df_samples <- df_in %>%
  dplyr::filter(Sample != "background") %>%
  dplyr::filter(Sample %in% metadata$Sample) %>%
  mutate(Sample = as.numeric(Sample))

# Subtract background, set negative values to zero
df_samples[,analyte_range] <- sweep(df_samples[,analyte_range],2,df_b)
df_samples[df_samples<0] <- 0 # analytes below background set to zero

glimpse(df_samples)
```



#### Create unicode labels for greek characters

To enable the use of greek characters in figures (created in the main analysis script) without complicating the column names of our sample data frame, create a "greekversion" array to store unicode labels.

```{r greek}
# create unicode labels for greek characters
greek = colnames(df_samples)
greek[which(greek=="AB40")] = "A\u03b240"
greek[which(greek=="AB42")] = "A\u03b242"
greek[which(greek=="AB42/40")] = "A\u03b242/40"
greek[which(greek=="tTau")] = "Total Tau"
greek[which(greek=="pTau")] = "Phospho-Tau T181"
greek[which(greek=="IFN-g")] = "IFN-\u03b3"
greek[which(greek=="IL-1a")] = "IL-1\u03b1"
greek[which(greek=="IL-1b")] = "IL-1\u03b2"
greek[which(greek=="MIP-1a")] = "MIP-1\u03b1"
greek[which(greek=="MIP-1b")] = "MIP-1\u03b2"
greek[which(greek=="TNF-a")] = "TNF-\u03b1"

greek[analyte_range]
```

#### Create color bars for use in figures

For analyses that maintain the sample order used in df_samples, we create a color bar for injury number and time point after injury. These will be used to denote experimental groups in bar plots, heatmaps, and other visualizations.

```{r colorbars, fig.height=3.5, fig.width=5}
# create color bars for use in figures
InjPalette = rev(RColorBrewer::brewer.pal(6,"Spectral"))
TimePalette = RColorBrewer::brewer.pal(4,"Purples")

InjColors = case_when(
  df_samples$Injury=="Sham" ~ InjPalette[1],
  df_samples$Injury== "1xCHI" ~ InjPalette[2],
  df_samples$Injury== "pre-3xCHI" ~ InjPalette[3],
  df_samples$Injury== "3xCHI" ~ InjPalette[4],
  df_samples$Injury== "pre-5xCHI" ~ InjPalette[5],
  df_samples$Injury== "5xCHI" ~ InjPalette[6])

TimeColors = case_when(
  df_samples$TimePoint== 0 ~ TimePalette[1],
  df_samples$TimePoint== 0.5 ~ TimePalette[2],
  df_samples$TimePoint== 4 ~ TimePalette[3],
  df_samples$TimePoint== 24 ~ TimePalette[4],)

barplot(rep(1,length(InjPalette)),col=InjPalette,names.arg = c("Sham","1xCHI","2xCHI","3xCHI","4xCHI","5xCHI"))
barplot(rep(1,length(TimePalette)),col=TimePalette,names.arg = c("Pre-Injury","30min","4hr","24hr"))
```

#### Save protein data objects

```{r protein_save}
save(df_samples,analyte_range,greek,InjPalette,TimePalette,InjColors,TimeColors,file="R Data/proteins.rda") #  save to access later
```

### Load and save gene sets for gene set variation analysis (GSVA)

Gene set variation analysis (GSVA) will be conducted on three collections: the C2 collection from the Molecular Signatures Database (MSigDB), custom curated annotations from astrocytes, and custom curated annotations from neurons.

MSigDB: <https://www.gsea-msigdb.org/gsea/msigdb/human/collections.jsp>
Custom Sets: <https://www.sciencedirect.com/science/article/pii/S0969996122000468>

```{r gsva_load, eval=FALSE}
# MSigDB c2
geneSetsIn=readxl::read_excel("Reference Files/c2.cp.v7.0.symbols.xlsx",col_names=FALSE) %>%
  dplyr::select(-2) %>%
  t() 
colnames(geneSetsIn) = geneSetsIn[1,]
gs_MSigDB_C2 = geneSetsIn[-1,] %>%
  as.data.frame() %>%
  as.list() %>%
  lapply(function(x) x[!is.na(x)])

# Astrocyte Annotations
gs_astro=readxl::read_excel("Reference Files/GeneSets_Astrocytes_FXN1.xlsx") %>%
  as.list() %>%
  lapply(function(x) x[!is.na(x)])

# Neuron Annotations
gs_neuron=readxl::read_excel("Reference Files/GeneSets_Neurons_FXN1.xlsx") %>%
  as.list() %>%
  lapply(function(x) x[!is.na(x)])

save(gs_MSigDB_C2,gs_astro,gs_neuron,file="R Data/gs.rda")
```
```{r load_gs, include=FALSE}
load("R Data/gs.rda")
```

```{r show_custom}
glimpse(gs_astro)
glimpse(gs_neuron)
```


### Normalization and variance stabilization with DESeq2

#### Setup and load counts
```{r DESeq2_setup}
rm(list=ls())   # clears "workspace"
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))  # tells R that the folder where this code resides is our working directory
load("R Data/genecounts.rda")
```

#### Normalization

DESeq2 uses "median of ratios" method, as described here: <https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html>

```{r DESeq2_norm, eval=FALSE}
# DESeq2 Normalization 
des_mat = count %>%
  round(digits=0) %>%
  as.matrix()
metadata$AbsoluteTime=factor(metadata$AbsoluteTime)
dds <- DESeq2::DESeqDataSetFromMatrix(countData = des_mat,
                              colData = metadata,
                              design = ~ AbsoluteTime)
dds <- DESeq2::DESeq(dds)
```

#### Variance stabilizing transform

VST is recommended for WGCNA: <https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html>

VST in DESeq2: <http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#variance-stabilizing-transformation>

```{r DESeq2_vst, eval=FALSE}
# Variance Stabilizing Transform
vsd <- DESeq2::getVarianceStabilizedData(dds)
datExpr = t(vsd)

cleanDat = vsd
rownames(cleanDat) = str_c(gene.df$geneName,"|",rownames(vsd))

save(datExpr,cleanDat,metadata,gene.df,file="R Data/geneVSD.rda")
```

